### PR TITLE
[PAC][llvm-readobj][AArch64][ELF] Support `GNU_PROPERTY_AARCH64_FEATURE_PAUTH`

### PR DESCRIPTION
**Pull request by kovdan01**
Reland #85231 after fixing build failure https://lab.llvm.org/buildbot/#/builders/186/builds/15631.
Use `PRIx64` for format output of `uint64_t` as hex.
Original PR description below.

This adds support for `GNU_PROPERTY_AARCH64_FEATURE_PAUTH` feature (as defined in https://github.com/ARM-software/abi-aa/pull/240) handling in llvm-readobj and llvm-readelf. The following constants for supported platforms are also introduced:

- `AARCH64_PAUTH_PLATFORM_INVALID = 0x0`
- `AARCH64_PAUTH_PLATFORM_BAREMETAL = 0x1`
- `AARCH64_PAUTH_PLATFORM_LLVM_LINUX = 0x10000002`

For the llvm_linux platform, output of the tools contains descriptions of PAuth features which are enabled/disabled depending on the version value. Version value bits correspond to the following `LangOptions` defined in #85232:

- bit 0: `PointerAuthIntrinsics`;
- bit 1: `PointerAuthCalls`;
- bit 2: `PointerAuthReturns`;
- bit 3: `PointerAuthAuthTraps`;
- bit 4: `PointerAuthVTPtrAddressDiscrimination`;
- bit 5: `PointerAuthVTPtrTypeDiscrimination`;
- bit 6: `PointerAuthInitFini`.

Support for `.note.AARCH64-PAUTH-ABI-tag` is dropped since it's deleted
from the spec in ARM-software/abi-aa#250.

### CODE REVIEW

**Review by MaskRay — APPROVED** [review]

File: llvm/tools/llvm-readobj/ELFDumper.cpp [hunk 1]
```diff
@@ -5105,6 +5106,73 @@ template <class ELFT> void GNUELFDumper<ELFT>::printAddrsig() {
...
+    Flags[AARCH64_PAUTH_PLATFORM_LLVM_LINUX_VERSION_VPTRTYPEDISCR] =
+        "VTPtrTypeDiscrimination";
+    Flags[AARCH64_PAUTH_PLATFORM_LLVM_LINUX_VERSION_INITFINI] = "InitFini";
+
+    static_assert(AARCH64_PAUTH_PLATFORM_LLVM_LINUX_VERSION_INITFINI ==
+                      AARCH64_PAUTH_PLATFORM_LLVM_LINUX_VERSION_LAST,
+                  "Update when new enum items are defined");
+
+    std::string Desc;
+    for (uint32_t I = 0; I < Flags.size(); ++I) {
```
Comment (MaskRay) [inline llvm/tools/llvm-readobj/ELFDumper.cpp:5159]: https://llvm.org/docs/CodingStandards.html#don-t-evaluate-end-every-time-through-a-loop

If you define `End` in the for loop, then you can avoid the resize(xx-2) below if `I == End-1`.

File: llvm/test/tools/llvm-readobj/ELF/AArch64/aarch64-feature-pauth.s [hunk 2]
```diff
@@ -1,98 +1,211 @@
...
-.long 4
-.long 24
-.long 1
-.asciz "ARM"
-
-.quad 42         // platform
-.quad 1          // version
-.quad 0x0123456789ABCDEF // extra data
+#--- gnu-42-1.s
+
```
Comment (MaskRay) [inline llvm/test/tools/llvm-readobj/ELF/AArch64/aarch64-feature-pauth.s:4]: the majority `#---` use cases don't place a blank line after `#---`.

File: llvm/tools/llvm-readobj/ELFDumper.cpp (hunk 1 above)
Comment (kovdan01) [reply llvm/tools/llvm-readobj/ELFDumper.cpp]: Fixed, thanks, see 3beb910bb9c9460b3127af94149fc0bb2f26e54d. I'm personally not sure though if adding an extra check against `I == End - 1` at every loop iteration is worth saving potential extra memory allocation in `std::string` due to unneeded 2 bytes - but if it's preferred, let is be so.

File: llvm/test/tools/llvm-readobj/ELF/AArch64/aarch64-feature-pauth.s (hunk 2 above)
Comment (kovdan01) [reply llvm/test/tools/llvm-readobj/ELF/AArch64/aarch64-feature-pauth.s]: Fixed, thanks, see 3beb910bb9c9460b3127af94149fc0bb2f26e54d

File: llvm/tools/llvm-readobj/ELFDumper.cpp (hunk 1 above)
Comment (MaskRay) [reply llvm/tools/llvm-readobj/ELFDumper.cpp]: Thanks. Alternatively, you can prepend ", " if `I != 0`, in case the code is shorter.

LGTM.

**Review by MaskRay — APPROVED** [review]

**Pull request merged**